provider/akamai/akamai.go

```diff
@@ -79,84 +86,126 @@ type akamaiZone struct {
+				edgeGridConfig.Debug = dbgval
+				log.Debugf("Edgegrid debug set to %s", envval)
+			}
+		}
 	}
 
 	provider := &AkamaiProvider{
 		domainFilter: akamaiConfig.DomainFilter,
 		zoneIDFilter: akamaiConfig.ZoneIDFilter,
-		config:       edgeGridConfig,
+		config:       &edgeGridConfig,
 		dryRun:       akamaiConfig.DryRun,
-		client:       &akamaiOpenClient{},
 	}
-	return provider
+	if akaService != nil {
+		log.Debugf("Using STUB")
+		provider.client = akaService
+	} else {
+		provider.client = provider
+	}
+
+	// Init library for direct endpoint calls
+	dns.Init(edgeGridConfig)
+
+	return provider, nil
 }
 
-func (p *AkamaiProvider) request(method, path string, body io.Reader) (*http.Response, error) {
-	req, err := p.client.NewRequest(p.config, method, fmt.Sprintf("https://%s/%s", p.config.Host, path), body)
-	if err != nil {
-		log.Errorf("Akamai client failed to prepare the request")
-		return nil, err
-	}
-	resp, err := p.client.Do(p.config, req)
+func (p AkamaiProvider) ListZones(queryArgs dns.ZoneListQueryArgs) (*dns.ZoneListResponse, error) {
+	return dns.ListZones(queryArgs)
+}
 
-	if err != nil {
-		log.Errorf("Akamai client failed to do the request")
-		return nil, err
-	}
-	if !c.IsSuccess(resp) {
-		return nil, c.NewAPIError(resp)
-	}
+func (p AkamaiProvider) GetRecordsets(zone string, queryArgs dns.RecordsetQueryArgs) (*dns.RecordSetResponse, error) {
+	return dns.GetRecordsets(zone, queryArgs)
+}
 
-	return resp, err
+func (p AkamaiProvider) CreateRecordsets(recordsets *dns.Recordsets, zone string, reclock bool) error {
+	return recordsets.Save(zone, reclock)
 }
 
-//Look here for endpoint documentation -> https://developer.akamai.com/api/web_performance/fast_dns_zone_management/v2.html#getzones
-func (p *AkamaiProvider) fetchZones() (zones akamaiZones, err error) {
-	log.Debugf("Trying to fetch zones from Akamai")
-	resp, err := p.request("GET", "config-dns/v2/zones?showAll=true&types=primary%2Csecondary", nil)
-	if err != nil {
-		log.Errorf("Failed to fetch zones from Akamai")
-		return zones, err
+func (p AkamaiProvider) GetRecord(zone string, name string, recordtype string) (*dns.RecordBody, error) {
+	return dns.GetRecord(zone, name, recordtype)
+}
+
+func (p AkamaiProvider) DeleteRecord(record *dns.RecordBody, zone string, recLock bool) error {
+	return record.Delete(zone, recLock)
+}
+
+func (p AkamaiProvider) UpdateRecord(record *dns.RecordBody, zone string, recLock bool) error {
+	return record.Update(zone, recLock)
+}
+
+// Fetch zones using Edgegrid DNS v2 API
+func (p AkamaiProvider) fetchZones() (akamaiZones, error) {
+	log.Debugf("Fetching Akamai Edge DNS zones")
+	filteredZones := akamaiZones{Zones: make([]akamaiZone, 0)}
+	queryArgs := dns.ZoneListQueryArgs{Types: "primary", ShowAll: true}
+	// filter based on contractIds
+	if len(p.zoneIDFilter.ZoneIDs) > 0 {
+		queryArgs.ContractIds = strings.Join(p.zoneIDFilter.ZoneIDs, ",")
 	}
+	resp, err := p.client.ListZones(queryArgs) // don't worry about paged results
```

Comment:
What does that mean `don't worry about paged results` 🙂 ?

Reply:
The API supports retrieval of all zones in one response or as a series of requests returning a pages, e.g. X zones at a time

Reply:
then maybe you should write that so it's more clear for someone which is using the provider and wants to look at the code

Reply:
Updated